File: dotnet/src/DocGen/Clients/SignaturesClient.cs
using Dokmatiq.DocGen.Internal;
using Dokmatiq.DocGen.Models;

namespace Dokmatiq.DocGen.Clients;

/// <summary>
/// Client for digital signature operations.
/// </summary>
public sealed class SignaturesClient
{
    private readonly Transport _transport;

    internal SignaturesClient(Transport transport) => _transport = transport;

    /// <summary>Upload a PKCS#12 certificate.</summary>
    public Dictionary<string, object> UploadCertificate(string filePath, string password)
        => _transport.Upload<Dictionary<string, object>>("/api/signatures/certificates", "file",
            FileUtils.ReadBytes(filePath), FileUtils.DetectFilename(filePath),
            new Dictionary<string, string> { ["password"] = password });

    /// <summary>List uploaded certificates.</summary>
    public List<Dictionary<string, object>> ListCertificates()
        => _transport.RequestList<Dictionary<string, object>>(HttpMethod.Get, "/api/signatures/certificates");

    /// <summary>Delete a certificate by name.</summary>
    public void DeleteCertificate(string name)
        => _transport.Delete($"/api/signatures/certificates/{name}");

    /// <summary>Get certificate details.</summary>
    public Dictionary<string, object> CertificateInfo(string name)
        => _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Get, $"/api/signatures/certificates/{name}");

    /// <summary>Digitally sign a PDF.</summary>
    public byte[] Sign(string filePath, string certificateName, string certificatePassword,
        string? reason = null, string? location = null, VisibleSignatureConfig? visibleSignature = null)
    {
        var body = new Dictionary<string, object?>
        {
            ["pdfBase64"] = FileUtils.ToBase64(filePath),
            ["certificateName"] = certificateName,
            ["certificatePassword"] = certificatePassword,
            ["reason"] = reason,
            ["location"] = location,
            ["visibleSignature"] = visibleSignature
        };
        return _transport.RequestBytes(HttpMethod.Post, "/api/signatures/sign", body);
    }

    /// <summary>Sign a PDF from bytes.</summary>
    public byte[] Sign(byte[] data, string certificateName, string certificatePassword)
        => _transport.RequestBytes(HttpMethod.Post, "/api/signatures/sign",
            new { pdfBase64 = FileUtils.ToBase64(data), certificateName, certificatePassword });

    /// <summary>Verify signatures in a PDF file.</summary>
    public SignatureVerifyResult Verify(string filePath)
        => _transport.RequestJson<SignatureVerifyResult>(HttpMethod.Post, "/api/signatures/verify",
            new { pdfBase64 = FileUtils.ToBase64(filePath) });

    /// <summary>Verify signatures from bytes.</summary>
    public SignatureVerifyResult Verify(byte[] data)
        => _transport.RequestJson<SignatureVerifyResult>(HttpMethod.Post, "/api/signatures/verify",
            new { pdfBase64 = FileUtils.ToBase64(data) });
}

File: java/src/main/java/com/dokmatiq/docgen/client/SignaturesClient.java
package com.dokmatiq.docgen.client;

import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.internal.Transport;
import com.dokmatiq.docgen.model.SignatureVerifyResult;
import com.dokmatiq.docgen.model.VisibleSignatureConfig;

import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Client for digital signature operations. */
public class SignaturesClient {
    private final Transport transport;

    public SignaturesClient(Transport transport) {
        this.transport = transport;
    }

    /** Upload a PKCS#12 certificate. */
    public Map<?, ?> uploadCertificate(Path file, String password) {
        return transport.upload("/api/signatures/certificates", "file",
                FileUtils.readBytes(file), FileUtils.detectFilename(file),
                Map.of("password", password), Map.class);
    }

    /** List uploaded certificates. */
    public List<Map> listCertificates() {
        return transport.requestList("GET", "/api/signatures/certificates", Map.class);
    }

    /** Delete a certificate by name. */
    public void deleteCertificate(String name) {
        transport.delete("/api/signatures/certificates/" + name);
    }

    /** Get certificate details. */
    public Map<?, ?> certificateInfo(String name) {
        return transport.requestJson("GET", "/api/signatures/certificates/" + name, null, Map.class);
    }

    /** Digitally sign a PDF. */
    public byte[] sign(Path file, String certificateName, String certificatePassword) {
        return sign(file, certificateName, certificatePassword, null, null, null);
    }

    /** Digitally sign a PDF with options. */
    public byte[] sign(Path file, String certificateName, String certificatePassword,
                       String reason, String location, VisibleSignatureConfig visibleSignature) {
        var body = new HashMap<String, Object>();
        body.put("pdfBase64", FileUtils.toBase64(file));
        body.put("certificateName", certificateName);
        body.put("certificatePassword", certificatePassword);
        if (reason != null) body.put("reason", reason);
        if (location != null) body.put("location", location);
        if (visibleSignature != null) body.put("visibleSignature", visibleSignature);
        return transport.requestBytes("POST", "/api/signatures/sign", body);
    }

    /** Sign a PDF from bytes. */
    public byte[] sign(byte[] data, String certificateName, String certificatePassword) {
        return transport.requestBytes("POST", "/api/signatures/sign",
                Map.of("pdfBase64", FileUtils.toBase64(data),
                        "certificateName", certificateName,
                        "certificatePassword", certificatePassword));
    }

    /** Verify signatures in a PDF. */
    public SignatureVerifyResult verify(Path file) {
        return transport.requestJson("POST", "/api/signatures/verify",
                Map.of("pdfBase64", FileUtils.toBase64(file)), SignatureVerifyResult.class);
    }

    /** Verify signatures from bytes. */
    public SignatureVerifyResult verify(byte[] data) {
        return transport.requestJson("POST", "/api/signatures/verify",
                Map.of("pdfBase64", FileUtils.toBase64(data)), SignatureVerifyResult.class);
    }
}
